src/server/game/Movement/MotionMaster.h

```diff
@@ -114,31 +136,67 @@ class TC_GAME_API MotionMaster
 
         void LaunchMoveSpline(Movement::MoveSplineInit&& init, uint32 id = 0, MovementSlot slot = MOTION_SLOT_ACTIVE, MovementGeneratorType type = EFFECT_MOTION_TYPE);
     private:
-        typedef std::vector<MovementGenerator*> MovementList;
-
-        void pop();
 
-        bool NeedInitTop() const;
-        void InitTop();
+        void AddFlag(uint8 const flag) { _flags |= flag; }
+        bool HasFlag(uint8 const flag) const { return (_flags & flag) != 0; }
+        void RemoveFlag(uint8 const flag) { _flags &= ~flag; }
 
-        void Mutate(MovementGenerator* m, MovementSlot slot);
+        MovementGeneratorPointer const* Top() const;
+        MovementGeneratorPointer const* Pop();
+        void DirectClear(bool reset);
+        void DirectAdd(MovementGenerator* movement, MovementSlot slot);
+        void DirectRemove(MovementGeneratorPointer const* movement, MovementSlot slot);
 
-        void DirectClean(bool reset);
-        void DelayedClean();
-        void DirectClean(MovementSlot slot);
-        void DelayedClean(MovementSlot slot);
-        void DirectExpire(bool reset);
-        void DelayedExpire();
-        void DirectDelete(MovementGenerator* curr);
-        void DelayedDelete(MovementGenerator* curr);
-        void ClearExpireList();
+        void Delete(MovementGeneratorPointer const* movement, bool active, bool movementInform);
+        void DeleteDefault(bool active, bool movementInform);
+        void UpdateBaseUnitStates(MovementGenerator* movement);
+        void ClearBaseUnitStates();
 
-        MovementGenerator* _slot[MAX_MOTION_SLOT];
-        bool _initialize[MAX_MOTION_SLOT];
-        MovementList _expireList;
         Unit* _owner;
-        int _top;
-        uint8 _cleanFlag;
+        MovementGeneratorPointer _defaultGenerator;
+        std::multiset<MovementGeneratorPointer, MovementGeneratorComparator> _generators;
+        std::unordered_multimap<uint32, MovementGenerator const*> _baseUnitStatesMap;
+        std::deque<std::shared_ptr<MotionMasterDelayedAction>> _delayedActions;
```

Comment:
shouldn't this be unique_ptr instead?

Reply:
well now that u say it... since its never _shared_ ... makes sense 💃